{{refimprove|time=2013-03-19T17:46:05+00:00}}
{{expand|time=2009-12-25T13:31:32+00:00}}
{{noteTA
|G1=IT
}}
'''异常处理'''，是[[编程语言|编程语言]]或计算机硬件里的一种机制，用于处理软件或信息系统中出现的异常状况（即超出程序正常执行流程的某些特殊条件）。

各种编程语言在处理异常方面具有非常显著的不同点（错误检测与异常处理区别在于：错误检测是在正常的程序流中，处理不可预见问题的代码，例如一个调用操作未能成功结束）。某些编程语言有这样的函数：当输入存在非法数据时不能被安全地调用，或者返回值不能与异常进行有效的区别。例如，C语言中的atoi函数（ASCII串到整数的转换）在输入非法时可以返回0。在这种情况下编程者需要另外进行错误检测（可能通过某些辅助全局变量如C的errno），或进行输入检验（如通过正则表达式），或者共同使用这两种方法。

通过异常处理，我们可以对用户在程序中的非法输入进行控制和提示，以防程序崩溃。

从进程的视角，硬件中断相当于可恢复异常，虽然中断一般与程序流本身无关。

从子程序编程者的视角，异常是很有用的一种机制，用于通知外界该子程序不能正常执行。如输入的数据无效（例如除数是0），或所需资源不可用（例如文件丢失）。如果系统没有异常机制，则编程者需要用返回值来标示发生了哪些错误。

== 硬件中的异常处理 ==

硬件的异常处理机制由 CPU 完成。这种机制支持错误检测，在发生错误后会将程序流跳转到专门的错误处理过程（{{lang-en|error handling routines}}）中。发生异常前的状态存储在栈上。<ref>{{cite web
 | url         = http://processors.wiki.ti.com/
 | title         = Hardware Exceptions Detection
 | date         = 2011-11-24
 | publisher         = TEXAS INSTRUMENTS
 | archiveurl         = http://wayback.vefsafn.is/wayback/20131110061200/http%3A//processors.wiki.ti.com/index.php/Main_Page
 | archivedate         = 2013-11-10
 | quote         = 
 | accessdate         = 2012-10-05
 | deadurl         = yes
 | df         = 
 | language = en
 }}</ref>

== 操作系统提供的异常处理设施 ==

针对程序中可能发生的异常，[[操作系统|操作系统]]可能通过 [[进程间通讯|IPC]] 来提供对应的处理设施。进程执行过程中发生的中断通常由操作提供的「中断服务子程序」处理，操作系统可以藉此向该进程发送[[Unix信号|信号]]。进程可以通过注册信号处理器的方式自行处理信号，也可以让操作系统执行默认行为（比如终止该程序）。

== 软件中的异常处理 ==

[[编程语言|编程语言]]的异常处理机制中，'''异常'''（{{lang-en|exception}}）这一术语所描述的通常是一种数据结构，这种数据结构可以存储与某种异常情况（{{lang-en|exceptional condition}}）相关的信息。'''抛'''（{{lang-en|throw}}）是用来移交控制权的机制，抛出异常也可以称作'''引发'''（{{lang-en|raise}}）异常。异常抛出后，控制权会被移交至某处的'''接'''（{{lang-en|catch}}）。

从[[子程序|子程序]]作者的角度看，如果要表示当前子程序无法正常执行，抛出异常是很好的选择。无法正常执行的原因可以是输入参数无效（比如值在函数的[[定义域|定义域]]之外），也可以是无法获得所需的资源（比如文件不存在、硬盘出错、内存不足）等等。在不支持异常的系统中，子程序需要通过返回特殊的{{Tsl|en|Error code|错误码}}实现类似的功能。然而返回错误码可能导致{{Tsl|en|Semipredicate problem|不完全预测问题}}，子程序的使用方需要编写额外的代码，才能将普通的返回值与错误码相区别。

==异常安全==
一段代码是'''异常安全的'''，如果这段代码运行时的失败不会产生有害后果，如[[内存泄露|内存泄露]]、存储数据混淆、或无效的输出。异常安全可分成不同层次：
# '''失败透明（failure transparency）'''，也称作'''不抛出保证（no throw guarantee）'''：代码的运行保证能成功并满足所有的约束条件，即使存在异常情况。如果出现了异常，将不会对外进一步抛出该异常。（异常安全的最好的层次）
# '''提交或卷回的语义（commit or rollback semantics）'''，或称作'''强异常安全（strong exception safety）'''或'''无变化保证（no-change guarantee）'''：运行可以是失败，但失败的运行保证不会有负效应，因此所有涉及的数据都保持代码运行前的初始值。<ref>http://www.open-std.org/jtc1/sc22/wg21/docs/papers/1997/N1077.asc</ref>
# '''基本异常安全（basic exception safety）'''：失败运行的已执行的操作可能引起了副作用，但会保证状态不变。所有存储数据保持有效值，即使这些数据与异常发生前的值有所不同。
# '''最小异常安全（minimal exception safety）'''也称作'''无泄漏保证（no-leak guarantee）'''：失败运行的已执行的操作可能在存储数据中保存了无效的值，但不会引起崩溃，资源不会泄漏。
# '''异常不安全（no exception safety）''':没有保证（最差的异常安全层次）。

例如，考虑一个smart vector类型，如C++'s <code>std::vector</code>或Java's <code>ArrayList</code>。当一个数据项<code>x</code>插入vector <code>v</code>，必须实际增加<code>x</code>的值到vector的内部对象列表中并且修改vector的计数域以正确表示<code>v</code>中保存了多少数据项；此时如果已有的存储空间不够大，就需要分配新的内存。内存分配可能会失败并抛出异常。因此，vector数据类型如果是“失败透明”保证将会非常困难甚至不可能实现。但vector类型提供“强异常安全”保证却是相当容易的；在这种情况下，<code>x</code>插入<code>v</code>或者成功，或者<code>v</code>保持不变。如果vector类型仅提供“基本异常安全”保证，如果数据插入失败，<code>v</code>可能包含也可能不包含<code>x</code>的值，但至少<code>v</code>的内部表示是一致的。但如果vector数据类型是“最小异常安全”保证，<code>v</code>可能会是无效的，例如<code>v</code>的计数域被增加了，但<code>x</code>并未实际插入，使得内部状态不一致。对于“异常不安全”的实现，程序可能会崩溃，例如写入数据到无效的内存。

通常至少需要基本异常安全。失败透明是难于实现的，特别是在编写库函数时，因为对应用程序的复杂知识缺少获知。

==程序设计语言的异常机制==
许多常见的程序设计语言，包括[[Actionscript|Actionscript]]，[[Ada程序设计语言|Ada]]，[[BlitzMax|BlitzMax]]，[[C++|C++]]，[[C_Sharp|C#]]，[[D程序设计语言|D]]，[[ECMAScript|ECMAScript]]，[[Eiffel|Eiffel]]，[[Java程序设计语言|Java]]，[[ML程序设计语言|ML]]，[[Object_Pascal|Object Pascal]]（如[[Delphi|Delphi]]，[[Free_Pascal|Free Pascal]]等），[[Objective-C|Objective-C]]，[[Ocaml|Ocaml]]，[[PHP|PHP]]（version 5），[[PL/1|PL/1]]，[[Prolog|Prolog]]，[[Python|Python]]，[[REALbasic|REALbasic]]，[[Ruby|Ruby]]，[[Visual_Prolog|Visual Prolog]]以及大多数[[.NET_Framework|.NET]]程序设计语言，内建的异常机制都是沿着函数-{zh-hans:调用栈;zh-hant:呼叫堆疊}-的函数调用逆向搜索，直到遇到异常处理代码为止。一般在这个异常处理代码的搜索过程中逐级完成[[堆疊輾轉開解|-{zh-hans:栈卷回;zh-hant:堆疊輾轉開解;}-]]（stack unwinding）。但[[Common_Lisp|Common Lisp]]是个例外，它不采取-{zh-hans:栈卷回;zh-hant:堆疊輾轉開解;}-，因此允许异常处理完后在抛出异常的代码处原地恢复执行。而 [[Visual_Basic|Visual Basic]]（尤其是在其早于 .net 的版本，例如 6.0 中）走得更远：<code>on error</code> 语句可轻易指定发生异常后是重试（<code>resume</code>）还是跳过（<code>resume next</code>）还是执行程序员定义的错误处理程序（<code>goto ***</code>）。

多数语言的异常机制的语法是类似的：用<code>throw</code>或<code>raise</code>抛出一个异常对象（Java或C++等）或一个特殊可扩展的枚举类型的值（如Ada语言）；异常处理代码的作用范围用标记子句（<code>try</code>或<code>begin</code>开始的语言作用域）标示其起始，以第一个异常处理子句（<code>catch, except, rescue</code>等）标示其结束；可连续出现若干个异常处理子句，每个处理特定类型的异常。某些语言允许<code>else</code>子句，用于无异常出现的情况。更多见的是<code>finally, ensure</code>子句，无论是否出现异常它都将执行，用于释放异常处理所需的一些资源。

[[C++异常处理|C++异常处理]]是[[资源获取即初始化|资源获取即初始化]]（Resource-Acquisition-Is-Initialization）的基础。

C语言一般认为是不支持异常处理的。Perl语言可选择支持[[结构化异常处理|结构化异常处理]]（structured exception handling）。

Python语言对异常处理机制是非常普遍深入的，所以想写出不含<code>try, except</code>的程序非常困难。

==参考文献==
<references/>

{{数据类型}}

[[Category:控制流程|Category:控制流程]]
[[Category:軟體工程|Category:軟體工程]]